### PR TITLE
Add caching to the months dropdown filter for posts

### PR DESCRIPTION
**Pull request by WPprodigy**
## Description
Follow-up to https://github.com/Automattic/vip-go-mu-plugins/pull/1883.

tl;dr, the query to generate the year/month filter dropdown on posts pages can be a bit expensive on large sites and is never cached. There is a new filter in WP core that will allow us to intercept and cache the results: https://core.trac.wordpress.org/ticket/51660

### Performance: Cache the year/month dropdown on admin post pages

To prevent an expensive query occurring on every admin posts-edit page load, this caches the available months/years for the dropdown filter.

## Checklist

Please make sure the items below have been covered before requesting a review:

- [x] This change works and has been tested locally (or has an appropriate fallback).
- [x] This change works and has been tested on a Go sandbox.
- [ ] This change has relevant unit tests (if applicable).
- [ ] This change has relevant documentation additions / updates (if applicable).
- [x] I've created a changelog description that aligns with the provided examples.

## Steps to Test

1. Go to a listing page e.g.: `/wp-admin/edit.php` or a custom post type listing page (`/wp-admin/edit.php?post_type=myCustomPostType`) or media library (`/wp-admin/upload.php`)
2. When the page is first loaded, the query should take place and the cache is populated.
3. Reload the listing page and note the query does not need to re-occur.
4. Bust the cache by added a post that belongs to a new month/year: `wp post generate --count=1 --post_type=post --post_date=2021-08-25`. Reload and check that the new month/year appears in the dropdown filter.

### CODE REVIEW

**Comment by WPprodigy** [comment]
And to summarize the previous PR's feedback / what I plan to look into:

1) Switch to wp_cache_* API instead of transients.
2) Look into new cache invalidation strategy (only invalidate when really needed)
3) Perhaps un-duplicate the logic that exists in `wpcom_vip_media_library_months_with_files()` already.

**Comment by WPprodigy** [comment]
Made some adjustments:

- Both media library and posts dropdown are handled the same way.
- They cache the results for 24hours, and only invalidate when a new post w/ a "fresh" month is added.
- Things got messy quickly when dealing w/ filtered post statuses, so opted for just avoiding the cache entirely if somebody selects the "trash" post status filter for example.

File: performance/vip-tweaks.php [hunk 1]
```diff
@@ -25,50 +39,118 @@ function wpcom_vip_disable_performance_tweaks() {
...
+	foreach ( $existing_months as $date ) {
+		if ( $post_year === $date->year && $post_month === $date->month ) {
+			$found = true;
 		}
+	}
 
-		return $months;
+	if ( ! $found ) {
+		// The month/year doesn't exist yet, so just bust the cache and let it re-generate.
+		wp_cache_delete( $cache_key, 'vip' );
```
Comment (WPprodigy) [inline performance/vip-tweaks.php:154]: Could technically just add the new month to the cache directly, but I figured busting is okay here since we are already leaning on the 24hour TTL for "cleanup". It can't be that often that an entirely new month is added under normal editorial conditions, and even if somehow lots of new months are added (like an import or something) then the cache will just be busted the first time and not each import. So editorial flow wouldn't be interrupted or notice anything.

File: performance/vip-tweaks.php [hunk 2]
```diff
@@ -25,50 +39,118 @@ function wpcom_vip_disable_performance_tweaks() {
...
+		return false;
+	}
+
+	$post_year  = get_the_time( 'Y', $post_id );
+	$post_month = get_the_time( 'n', $post_id );
+
+	$found = false;
+	foreach ( $existing_months as $date ) {
+		if ( $post_year === $date->year && $post_month === $date->month ) {
+			$found = true;
```
Comment (jesse-greathouse) [inline performance/vip-tweaks.php:148]: I think there should be a `break;` here. I see no reason for the the loop to continue after a match is found. Do you?

**Review by jesse-greathouse — CHANGES_REQUESTED** [review]
Looks good. I did find a small suggestion that I think may possibly help.

File: performance/vip-tweaks.php (hunk 2 above)
Comment (WPprodigy) [reply performance/vip-tweaks.php:148]: Nice, good find.

File: performance/vip-tweaks.php [hunk 3]
```diff
@@ -15,8 +16,21 @@ function wpcom_vip_enable_performance_tweaks() {
 	if ( function_exists( 'wpcom_vip_enable_maybe_skip_old_slug_redirect' ) ) {
 		wpcom_vip_enable_maybe_skip_old_slug_redirect();
 	}
+
+	if ( is_admin() ) {
+		// Prevent expensive media library queries, and cache the available months for filtering.
+		add_filter( 'media_library_show_video_playlist', '__return_true' );
```
Comment (rebeccahum) [inline performance/vip-tweaks.php:22]: Do we still need to force return true if the default value is already true? https://github.com/WordPress/WordPress/blob/39bff93b6b274a873158a362725b1338ab1400e5/wp-includes/media.php#L4323

File: performance/vip-tweaks.php [hunk 4]
```diff
@@ -15,8 +16,21 @@ function wpcom_vip_enable_performance_tweaks() {
...
+		add_filter( 'media_library_show_audio_playlist', '__return_true' );
```
Comment (rebeccahum) [inline performance/vip-tweaks.php:23]: Same as above ^^.

File: performance/vip-tweaks.php [hunk 5]
```diff
@@ -25,50 +39,118 @@ function wpcom_vip_disable_performance_tweaks() {
...
-					FROM $wpdb->posts
-					WHERE post_type = %s
-					ORDER BY post_date DESC
-				", 'attachment' ) );
-			set_transient( 'wpcom_media_months_array', $months );
+/**
+ * Helper method to bust the "available months" cache when needed.
+ *
+ * Note that we only bust the cache when a new month is added,
+ * For deletions, we just let the 24-hour TTL handle that.
```
Comment (rebeccahum) [inline performance/vip-tweaks.php:127]: Why did you choose this approach over having no cache expiry and invalidating cache on post deletion?

Comment (rebeccahum) [inline performance/vip-tweaks.php:127]: Why did you choose this approach over having no cache expiry and invaliding on post deletion?

**Review by rebeccahum — CHANGES_REQUESTED** [review]

File: performance/vip-tweaks.php (hunk 3 above)
Comment (WPprodigy) [reply performance/vip-tweaks.php:22]: Nice find, looks like we can remove these now since WP 4.8 made it the default. These filters predate this PR, I just moved them up a few lines.

File: performance/vip-tweaks.php (hunk 5 above)
Comment (WPprodigy) [reply performance/vip-tweaks.php:127]: Most post deletions are quite unlikely to result in any change to the "available months". It would have to be the last post within that month to result in a change. Worst case is that the "last post" is deleted and somebody filters by a month within the same day and finds no posts there, but truly unlikely. Vs the high likelihood of deletes being fairly often and resulting in cache busts.

**Review by rebeccahum — APPROVED** [review]

**Comment by rebeccahum** [comment]
v20211215.0
